flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/SubtaskGatewayImpl.java

```diff
@@ -72,19 +79,28 @@
         FutureUtils.assertNoException(
                 result.handleAsync(
                         (success, failure) -> {
-                            if (failure != null && subtaskAccess.isStillRunning()) {
-                                String msg =
-                                        String.format(
-                                                EVENT_LOSS_ERROR_MESSAGE,
-                                                evt,
-                                                subtaskAccess.subtaskName());
-                                subtaskAccess.triggerTaskFailover(new FlinkException(msg, failure));
+                            if (failure != null) {
+                                if (subtaskAccess.isStillRunning()) {
+                                    String msg =
+                                            String.format(
+                                                    EVENT_LOSS_ERROR_MESSAGE,
+                                                    evt,
+                                                    subtaskAccess.subtaskName());
+                                    subtaskAccess.triggerTaskFailover(
+                                            new FlinkException(msg, failure));
+                                }
+
+                                nonSuccessFuturesTrack.removeFailedFuture(result);
```

Comment:
Hi Till, very thanks for the review! First for this issue, my initial concern for this method is that it seems might cause deadlocks:
1. For a checkpoint when it gets to the `completeCheckpointOnceEventsAreDone`, it would block the main thread and waits for all the pending event futures (with this method it would be the `whenAsync` one) are done. 
2. When the event sending result future finished, the thread finish it would also try to submit the `whenAsync` stage to the main thread, which would not get executed since the main thread is blocked.

Reply:
Hmm, let me check whether this is the case. If the checkpoint blocks the main thread, then this is another serious problem that should not happen.

Reply:
It doesn't block the main thread. It just creates a conjunct future of all pending event futures and chains the checkpoint completion to that future. That is all fully async.

Reply:
Yes, indeed, very sorry for I misunderstood at the original implementation.